### PR TITLE
 lib: lte_lc: add tests for public API using Unity/CMock 

### PR DESCRIPTION
**Pull request by maxd-nordic**
This patch adds a test suite to test the public API of the lte_lc
library. All public functions should be covered, although some
parts that depend on threads remain untested.

Signed-off-by: Maximilian Deubel <maximilian.deubel@nordicsemi.no>

### CODE REVIEW

File: tests/lib/lte_lc_api/src/lte_lc_api_test.c
Comment (tokangas) [inline tests/lib/lte_lc_api/src/lte_lc_api_test.c:298]:
```suggestion
void test_lte_lc_register_handler_twice(void)
```

File: tests/lib/lte_lc_api/src/lte_lc_api_test.c
Comment (tokangas) [inline tests/lib/lte_lc_api/src/lte_lc_api_test.c:337]: This and the following test case (`test_lte_lc_init_and_connect_home()`) don't actually connect to the network, because the response indicates that modem is already registered. This may be intentional, but maybe it could be visible in the test case name or in a comment.

**Review by gopiotr — COMMENTED** [review]
Please add also register into CODEOWNERS file with path to test `/tests/lib/lte_lc_api/` and person/people responsible for this test:
https://github.com/nrfconnect/sdk-nrf/blob/main/CODEOWNERS

Or if the same people are responsible for tests in `/tests/lib/lte_lc/`, then you can change this one register by add `*` sign like: `/tests/lib/lte_lc*/`

File: tests/lib/lte_lc_api/src/lte_lc_api_test.c
Comment (maxd-nordic) [reply tests/lib/lte_lc_api/src/lte_lc_api_test.c]: done

**Comment by maxd-nordic** [comment]
> Please add also register into CODEOWNERS file with path to test `/tests/lib/lte_lc_api/` and person/people responsible for this test: https://github.com/nrfconnect/sdk-nrf/blob/main/CODEOWNERS
> 
> Or if the same people are responsible for tests in `/tests/lib/lte_lc/`, then you can change this one register by add `*` sign like: `/tests/lib/lte_lc*/`

Added myself as a code owner. Does anyone else want to be on that list? @balaji-nordic @simensrostad

File: tests/lib/lte_lc_api/CMakeLists.txt
Comment (balaji-nordic) [inline tests/lib/lte_lc_api/CMakeLists.txt:17]: I dont think this documentation adds value. But no strong opinion.

File: tests/lib/lte_lc_api/CMakeLists.txt
Comment (balaji-nordic) [inline tests/lib/lte_lc_api/CMakeLists.txt:20]:
```suggestion
zephyr_include_directories(${NRF_DIR}/lib/lte_link_control/)
```

File: tests/lib/lte_lc_api/src/functions_to_mock.h
Comment (balaji-nordic) [inline tests/lib/lte_lc_api/src/functions_to_mock.h:4]: We should avoid creating this new file. Lets try this together when I am back in the office on Friday. :-)

**Review by balaji-nordic — CHANGES_REQUESTED** [review]
As commented elsewhere, lets make an attempt to get rid of the functions_to_mock.h and then proceed with the review. I am just suspecting that there may be some changes to the test implementation depending on how we do it. So marking this as 'Request changes' for now.

**Comment by balaji-nordic** [comment]
> > Please add also register into CODEOWNERS file with path to test `/tests/lib/lte_lc_api/` and person/people responsible for this test: https://github.com/nrfconnect/sdk-nrf/blob/main/CODEOWNERS
> > Or if the same people are responsible for tests in `/tests/lib/lte_lc/`, then you can change this one register by add `*` sign like: `/tests/lib/lte_lc*/`
> 
> Added myself as a code owner. Does anyone else want to be on that list? @balaji-nordic @simensrostad

I dont have to be on the list :)

File: tests/lib/lte_lc_api/src/lte_lc_api_test.c
Comment (trantanen) [inline tests/lib/lte_lc_api/src/lte_lc_api_test.c:52]: This is a general comment. I was hoping that we could create a general mock for nrf_modem_at_scanf, and also nrf_modem_at_printf (required when variadic parameters are used) so various tests could use it.
Have you discussed this with @lemrey?

**Comment by maxd-nordic** [comment]
Sorry for pinging all of you - this was a rebase mistake.

File: tests/lib/lte_lc_api/src/lte_lc_api_test.c
Comment (maxd-nordic) [reply tests/lib/lte_lc_api/src/lte_lc_api_test.c]: It could be nice to have a common mock for these! Unfortunately, this would get rather large since it would need to cover all of the different use cases along with possible outputs.

Comment (trantanen) [reply tests/lib/lte_lc_api/src/lte_lc_api_test.c]: I've tried to make some draft of how the generic solution could potentially look like so that other tests could utilize it also:
https://github.com/trantanen/sdk-nrf/blob/at_scanf_cmock/tests/lib/lte_lc_api/src/lte_lc_api_test.c#L288
We would need to put this into a library, e.g., test_mock_nrf_modem_at().
I didn't have time today to complete all tests with the mock trial but haven't yet found a reason why this wouldn't work. And so far I think this would be better as it would be more generic. I can continue tomorrow to make all tests pass with this.
I don't know if @lemrey has an idea how this would be even more generic.
Obviously, the ultimate solution would be to make variadic argument support into CMock itself.

Comment (glarsennordic) [reply tests/lib/lte_lc_api/src/lte_lc_api_test.c]: We in portland have been talking about the difficulty in mocking our variadic nrf_modem functions too. I actually brought up creating a centralized solution on one of Noric's testing Teams channels quite a while ago, but it didn't get any response, so the Portland team decided to pursue it independently. While we've not built anything yet, we do have a ticket describing a possible generalized solution: https://projecttools.nordicsemi.no/jira/browse/IRIS-4770

Tagging @plskeggs and @tony-le-24 on this

Comment (trantanen) [reply tests/lib/lte_lc_api/src/lte_lc_api_test.c]: Generic CMock for nrf_modem_at in PR #8766 has progressed and is fairly usable already. We just need a decision on which way to go and we should make that decision before merging these PRs.

**Comment by maxd-nordic** [comment]
This is now superseeded by https://github.com/nrfconnect/sdk-nrf/pull/8766

**Pull request closed without merging**
